### PR TITLE
feat!: prover input run on native arch

### PR DESCRIPTION
**Pull request by antoniolocascio**
## What ❔
This PR changes the way we compute prover input. Before, we needed to run the slow risc-v simulator to do so.
Now, we can do on native architecture.
This run is also responsible for computing the pubdata.

## Is this a breaking change?
- [ ] Yes
- [ ] No

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted.

### CODE REVIEW

File: basic_system/src/system_functions/modexp/mod.rs [hunk 1]
```diff
@@ -221,12 +229,23 @@ fn modexp_as_system_function_inner<
     );
 
     #[cfg(not(any(all(target_arch = "riscv32", feature = "proving"), test)))]
-    let output = ::modexp::modexp(
-        base.as_slice(),
-        exponent.as_slice(),
-        modulus.as_slice(),
-        allocator,
-    );
+    let output = if USE_DELEGATION {
```
Comment (AntonD3) [inline basic_system/src/system_functions/modexp/mod.rs]: Can we run delegation implementation depending on proving feature, so we can remove this extra argument?

And actually maybe it makes sense to rename delegation implementation to something like "proving implementation"/"advice_implementation"? As I understand it uses delegation depending on target

File: forward_system/src/run/output.rs [hunk 2]
```diff
@@ -134,12 +135,21 @@ impl<TR: TxResultCallback> From<ForwardRunningResultKeeper<TR>> for BlockOutput
             storage_writes,
             account_diffs,
             published_preimages,
-            pubdata,
+            // TODO: should we define 2 types of output?
```
Comment (AntonD3) [inline forward_system/src/run/output.rs]: I would remove pubdata from block output, and for now return pubdata as separate value(not in block output). In future we'll probably have `BatchOutput`

File: proof_running_system/src/system/bootloader.rs [hunk 3]
```diff
@@ -179,17 +178,13 @@ pub fn run_proving_inner<
     let _ = L::default().write_fmt(format_args!("IO implementer init is complete"));
 
     // Load all transactions from oracle and apply them.
-    let (mut oracle, public_input) = ProvingBootloader::<O, L>::run_prepared::<
-        BasicBootloaderProvingExecutionConfig,
-    >(oracle, &mut NopResultKeeper, &mut NopTracer::default())
-    .expect("Tried to prove a failing batch");
-
-    // disconnect oracle before returning, if some other post-logic is needed that doesn't use Oracle trait
```
Comment (AntonD3) [inline proof_running_system/src/system/bootloader.rs]: So, we don't need disconnect now?

**Review by AntonD3 — COMMENTED** [review]
Overall this looks great, left a few minor comments and questions

File: basic_system/src/system_functions/modexp/mod.rs (hunk 1 above)
Comment (antoniolocascio) [reply basic_system/src/system_functions/modexp/mod.rs]: I wanted to avoid using the feature, as we want to use the basic_system crate in both modes for the same binary. Otherwise, we'll need to somehow import the crate twice for server to run both the sequencing system and the prover input gen. 

The naming is just because it uses the modexp::delegation implementation, but can rename to advice, it makes more sense.

File: callable_oracles/src/lib.rs [hunk 4]
```diff
@@ -59,3 +59,23 @@ impl UsizeDeserializable for MemoryRegionDescriptionParams {
...
 }
+
+#[inline(always)]
+unsafe fn read_u64_words(ptr_u64: u64, len_words_u64: u64) -> Vec<u64> {
+    if ptr_u64 == 0 || len_words_u64 == 0 {
+        return vec![];
+    }
+    let addr = ptr_u64 as usize;
+    let len_words = len_words_u64 as usize;
+    core::slice::from_raw_parts(addr as *const u64, len_words).to_vec()
```
[the remark this answers is not in the document]
Comment (antoniolocascio) [reply callable_oracles/src/lib.rs]: Let me try, I agree

File: forward_system/src/run/output.rs (hunk 2 above)
Comment (antoniolocascio) [reply forward_system/src/run/output.rs]: Removed it, had to use a new branch for interface

File: basic_system/src/system_functions/mod.rs [hunk 5]
```diff
@@ -29,9 +29,9 @@ fn bytereverse(input: &mut [u8]) {
 /// No std system functions implementations.
 /// All of them are following EVM specs(for precompiles and keccak opcode).
 ///
-pub struct NoStdSystemFunctions;
+pub struct NoStdSystemFunctions<const USE_ADVICE: bool>;
```
Comment (AntonD3) [inline basic_system/src/system_functions/mod.rs]: nit: maybe worth to mentioned that `USE_ADVICE` affects only forward system

File: basic_system/src/system_functions/modexp/mod.rs [hunk 6]
```diff
@@ -25,29 +25,32 @@ pub const MODEXP_ADVICE_QUERY_ID: u32 = ADVICE_SUBSPACE_MASK | 0x10;
...
 #[derive(Debug, Default)]
-pub struct ModExpAdviceParams {
-    pub op: u32,          // Operation type (0 = division)
-    pub a_ptr: u32,       // Pointer to dividend
-    pub a_len: u32,       // Length of dividend in words
-    pub b_ptr: u32,       // Pointer to divisor
-    pub b_len: u32,       // Length of divisor in words
-    pub modulus_ptr: u32, // Pointer to modulus
-    pub modulus_len: u32, // Length of modulus in words
+pub struct ModExpAdviceParamsGeneric<W> {
```
Comment (AntonD3) [inline basic_system/src/system_functions/modexp/mod.rs]: can we just use usize?

File: callable_oracles/src/arithmetic/mod.rs [hunk 7]
```diff
@@ -1,12 +1,57 @@
...
+                } else {
+                    break;
+                }
+            }
+            &input[..digits]
+        }
+        let quotient = strip_leading_zeroes(&self.quotient);
+        let remainder = strip_leading_zeroes(&self.remainder);
+
+        // account for usize being u64 here
```
Comment (AntonD3) [inline callable_oracles/src/arithmetic/mod.rs]: nit: should we assert that usize indeed is 64 bit?

**Review by AntonD3 — COMMENTED** [review]
Overall looks good, there are some things that are not very clear for me around arithmetic advice(asked questions in comments), and please consider that it requires non-trivial integration from the platform side - pubdata is now not available from forward run

File: Cargo.toml [hunk 8]
```diff
@@ -61,8 +61,8 @@ keywords = ["blockchain", "zksync", "zk", "risc-v"]
 categories = ["cryptography"]
 
 [workspace.dependencies]
-zksync_os_evm_errors = { version = "0.0.14", default-features = false }
-zksync_os_interface = { version = "0.0.14", default-features = false }
+zksync_os_evm_errors = { git = "https://github.com/matter-labs/zksync-os-interface", branch = "alocascio-prover-input-gen", default-features =  false }
+zksync_os_interface = { git = "https://github.com/matter-labs/zksync-os-interface", branch = "alocascio-prover-input-gen" }
```
Comment (antoniolocascio) [inline Cargo.toml]: Thread to prevent merging without merging this first

**Comment by 0xVolosnikov** [comment]
@vv-dev-ai ptal

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `26ac689f4a9e226bfdac8ac3f8e97ee093d0dbf5` — **COMMENT** (4 inline comments).

File: callable_oracles/src/lib.rs [hunk 9]
```diff
@@ -60,12 +66,55 @@ impl UsizeDeserializable for MemoryRegionDescriptionParams {
...
+    let bytes = unsafe { core::slice::from_raw_parts(ptr, len_bytes) };
+    bytes.to_vec()
+}
+
+#[inline(always)]
+pub(crate) fn read_host_struct<T>(ptr_u64: u64) -> T {
+    let ptr = validate_host_pointer(ptr_u64, core::mem::align_of::<T>());
+
+    // Safety: the pointer was validated to be non-null and aligned for `T`.
+    unsafe { ptr.cast::<T>().read() }
```
Comment (vv-dev-ai) [inline callable_oracles/src/lib.rs:131]: [AI-review] **Medium:** The `// SAFETY:` comment is incomplete. `ptr.read()` requires not just non-null + aligned, but also that the pointed-to memory is initialized, valid for `size_of::<T>()` bytes, and that `T`'s validity invariant is satisfied by the bit pattern at that address. The pointer originates from the caller's own address space (stack-allocated `ModExpAdviceParams64` / `FieldOpsHint64`), so this is currently sound, but the safety comment should document these full preconditions.

Additionally, `read_host_struct` is generic over any `T` with no trait bounds. Adding `T: Copy` (or a custom marker trait) would prevent accidental use with types that have drop glue or non-trivial validity invariants.

Suggested safety comment:
```rust
// Safety: the caller guarantees that `ptr_u64` points to a fully
// initialized value of type `T` within the current process's
// address space. The pointer was validated to be non-null and
// aligned for `T`. The memory is not concurrently mutated.
```

File: callable_oracles/src/lib.rs [hunk 10]
```diff
@@ -60,12 +66,55 @@ impl UsizeDeserializable for MemoryRegionDescriptionParams {
...
+    if len_words_u64 == 0 {
+        return vec![];
+    }
+    let ptr = validate_host_pointer(ptr_u64, core::mem::align_of::<u64>());
+    let len_bytes = checked_byte_len(len_words_u64, core::mem::size_of::<u64>(), max_bytes);
+    let len_words = len_bytes / core::mem::size_of::<u64>();
+
+    // Safety: the caller supplied a non-null pointer aligned for `u64`, and
+    // `len_words` was derived from a checked byte-length computation.
+    let words = unsafe { core::slice::from_raw_parts(ptr.cast::<u64>(), len_words) };
```
Comment (vv-dev-ai) [inline callable_oracles/src/lib.rs:101]: [AI-review] **Low:** Same incomplete safety comment pattern as `read_host_struct`. The comment mentions non-null + alignment + bounds but omits the critical requirement that the memory region `[ptr, ptr + len_bytes)` must contain initialized data and remain valid for the duration of the read. This applies to `read_u8_words` at line 110 as well.

File: oracle_provider/src/lib.rs [hunk 11]
```diff
@@ -322,3 +328,85 @@ impl<M: MemorySource> NonDeterminismCSRSource<M> for ReadWitnessSource<M> {
...
+    where
+        I: UsizeSerializable + UsizeDeserializable,
+    {
+        let inner = self.original_source.raw_query(query_type, input)?;
+        // First add the length of the iterator.
+        let len = inner.len();
+        {
+            let mut read_items = self.read_items.borrow_mut();
+            // Len is multiplied by 2 to account for 32/64-bit mismatch
+            let len_u32 = u32::try_from(len * 2).expect("iterator length does not fit into u32");
```
Comment (vv-dev-ai) [inline oracle_provider/src/lib.rs]: [AI-review] **Low:** `len * 2` could overflow on a pathological oracle response before the `try_from` check catches it. Consider:
```rust
let len_u32 = u32::try_from(len.checked_mul(2).expect("response length overflow"))
    .expect("iterator length does not fit into u32");
```
This makes the multiplication overflow explicit rather than relying on `usize` wrapping (which wouldn't wrap on 64-bit in practice, but is cleaner).

File: basic_bootloader/src/bootloader/block_flow/zk/post_tx_op/post_tx_op_proving_singleblock_batch.rs [hunk 12]
```diff
@@ -246,9 +247,18 @@ where
                     state_diffs_hasher.update(value.current_value.as_u8_ref());
                 });
             let state_diffs_hash = state_diffs_hasher.finalize().into();
-            Ok((io.oracle, state_diffs_hash))
+
+            let _ = io
+                .oracle
+                .raw_query_with_empty_input(DISCONNECT_ORACLE_QUERY_ID)
+                .expect("must disconnect an oracle before performing arbitrary CSR access");
+            Ok((io.oracle, state_diffs_hash, batch_output))
```
Comment (vv-dev-ai) [inline basic_bootloader/src/bootloader/block_flow/zk/post_tx_op/post_tx_op_proving_singleblock_batch.rs]: [AI-review] **Nit:** The `.expect()` on `DISCONNECT_ORACLE_QUERY_ID` is consistent with the previous code in `proof_running_system`, but in proving mode a panic produces an undiagnosable abort. Since the `post_op` function already returns `Result`, consider propagating via `?`:
```rust
let _ = io.oracle
    .raw_query_with_empty_input(DISCONNECT_ORACLE_QUERY_ID)?;
```
Same pattern applies in `post_tx_op_proving_multiblock_batch.rs` and `post_tx_op_proving.rs`.

**Review by vv-dev-ai — COMMENTED** [review]
[AI-review] PR #418 — **COMMENT**

### Unsafe code documentation

The new host-memory reading helpers in `callable_oracles/src/lib.rs` (`read_host_struct`, `read_u64_words`, `read_u8_words`) have incomplete `// SAFETY:` comments. Per the codebase's unsafe code checklist, every unsafe block must document: pointer validity (points to live, initialized memory of the correct type), alignment, initialization, bounds, and lifetime (memory not freed during read). The current comments mention only non-null and alignment, omitting that the caller must guarantee the pointed-to memory is initialized and remains valid for the duration of the read.

All call sites are sound (pointers originate from stack-allocated structs within the same native process), but the documentation should be complete for future maintainers.

### Suggested Tests

- **Differential test: native vs RISC-V prover input for modexp with large inputs** — The modexp advice path has the most complex 32/64-bit adaptation (packed lengths, digit size conversion). A test with multi-limb modexp inputs (e.g., 2048-bit modulus) exercising `NativeArithmeticQuery` would strengthen coverage of the bigint digit size conversion logic. The existing `native_arithmetic_query_processes_valid_query` test uses only single-limb inputs.
- **Edge case: modexp with `q_len` or `r_len` that is odd in u32 words** — The native path asserts `q_len.is_multiple_of(2) && r_len.is_multiple_of(2)`. A test confirming this assertion fires when the oracle returns odd word counts would document this invariant.

File: basic_bootloader/src/bootloader/block_flow/zk/post_tx_op/post_tx_op_proving_singleblock_batch.rs (hunk 12 above)
Comment (antoniolocascio) [reply basic_bootloader/src/bootloader/block_flow/zk/post_tx_op/post_tx_op_proving_singleblock_batch.rs]: This moves in the next PR

File: basic_system/src/system_functions/modexp/advice/bigint.rs [hunk 13]
```diff
@@ -857,37 +856,79 @@ impl<'a, O: IOOracle> ModexpAdvisor for OracleAdvisor<'a, O> {
...
-                a_ptr: a_ptr.addr() as u32,
-                a_len: a_len as u32,
-                b_ptr: 0,
-                b_len: 0,
-                modulus_ptr: modulus_ptr.addr() as u32,
-                modulus_len: modulus_len as u32,
-            }
+        // We use different advice params depending on architecture
+        // Both are mostly the same, main difference is the width of pointers
+        #[cfg(target_arch = "riscv32")]
```
[the remark this answers is not in the document]
Comment (AntonD3) [reply basic_system/src/system_functions/modexp/advice/bigint.rs]: I understand that in practice we'll not run it on other 32-bit platform, but still technically this code is incorrect, maybe someone else will want to reuse this code, etc. Ofc, it's all weak arguments, but I saw we use `cfg(target_pointer_width = "64")` in other places, so why just not use it here?

File: crypto/src/bigint_delegation/u256.rs [hunk 14]
```diff
@@ -1,5 +1,5 @@
 use super::{delegation, DelegatedBarretParams, DelegatedModParams, DelegatedMontParams};
-use crate::ark_ff_delegation::{BigInt, BigInteger};
+use crate::{BigInt, BigInteger};
```
Comment (AntonD3) [inline crypto/src/bigint_delegation/u256.rs:2]: Isn't it supposed to use ark_ff_delegation BigInt?

File: callable_oracles/src/arithmetic/mod.rs (hunk 7 above)
Comment (AntonD3) [reply callable_oracles/src/arithmetic/mod.rs:37]: Same here. While it’s practically 99.9% fine, adding a small assertion helps maintain better code quality.

File: callable_oracles/src/arithmetic/mod.rs [hunk 15]
```diff
@@ -1,11 +1,60 @@
...
 use oracle_provider::OracleQueryProcessor;
 use risc_v_simulator::abstractions::memory::MemorySource;
 
 use crate::utils::{
     evaluate::{read_memory_as_u64, read_struct},
     usize_slice_iterator::UsizeSliceIteratorOwned,
 };
+use crate::{read_host_struct, read_u64_words};
+
+const HOST_MODEXP_ADVICE_MAX_BYTES: usize = 32 * 1024 * 1024;
```
Comment (AntonD3) [inline callable_oracles/src/arithmetic/mod.rs]: Where this const comes from?

File: crypto/src/bigint_delegation/u256.rs (hunk 14 above)
Comment (antoniolocascio) [reply crypto/src/bigint_delegation/u256.rs:2]: It is whenever we're in a proving environment, see lib.rs in crypto:
```
#[cfg(any(
    all(target_arch = "riscv32", feature = "bigint_ops"),
    feature = "proving",
    test
))]
pub use self::ark_ff_delegation::{BigInt, BigInteger};

#[cfg(not(any(
    all(target_arch = "riscv32", feature = "bigint_ops"),
    feature = "proving",
    test
)))]
pub use self::ark_ff::{BigInt, BigInteger};
```

File: callable_oracles/src/arithmetic/mod.rs (hunk 15 above)
Comment (antoniolocascio) [reply callable_oracles/src/arithmetic/mod.rs]: AI convinced me with the other uses cases of this limit, but actually it makes no sense to check this max bytes. This one I ignored as it's big enough for modexp in fusaka, but thinking more about it makes no sense to check this. Removed the checks altogether.

File: callable_oracles/src/arithmetic/mod.rs (hunk 7 above)
Comment (antoniolocascio) [reply callable_oracles/src/arithmetic/mod.rs:37]: Yeah, added a strong assert for this in the lib.rs of callable_oracles, so it should be enough

File: basic_system/src/system_functions/modexp/advice/bigint.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (antoniolocascio) [reply basic_system/src/system_functions/modexp/advice/bigint.rs]: Done, makes sense

File: basic_bootloader/src/bootloader/block_flow/ethereum/post_tx_op_proving.rs [hunk 16]
```diff
@@ -144,7 +145,12 @@ where
             &metadata.block_level.computed_header_hash
         );
 
-        Ok((io.oracle, metadata.block_level.computed_header_hash))
+        let _ = io
```
Comment (0xVolosnikov) [inline basic_bootloader/src/bootloader/block_flow/ethereum/post_tx_op_proving.rs]: Would be nice to use something more abstract - implement SimpleQuery or smth, to not touch internal implementation details here

File: callable_oracles/src/lib.rs [hunk 17]
```diff
@@ -60,12 +66,52 @@ impl UsizeDeserializable for MemoryRegionDescriptionParams {
     }
 }
 
+fn validate_host_pointer(ptr_u64: u64, alignment: usize) -> *const u8 {
+    assert!(ptr_u64 != 0);
```
Comment (0xVolosnikov) [inline callable_oracles/src/lib.rs:75]: Can you explain why?

**Comment by 0xVolosnikov** [comment]
@vv-dev-ai do an aggressive and paranoid review of this PR. This change is critical - you can use subagents to check it from different angles, with different context and different knowledge. Also take into account comments in this PR

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `c4c3a21069a2a6ad8f0b13fc7aa39efc934911cb` — **COMMENT** (3 inline comments).

**Review by vv-dev-ai — COMMENTED** [review]
[AI-review] PR #418 — **COMMENT**

### Findings

**1. `read_host_struct<T>` lacks `T: Copy` bound** (Medium, defense-in-depth)

The function is `pub(crate)` and accepts arbitrary `T`. All current callers pass `Copy` types (`ModExpAdviceParams64`, `FieldOpsHint64`), so this is sound today. But the API permits future callers to pass types with drop glue, which would create double-free UB via `ptr::read()`. Adding `T: Copy` is a zero-cost compile-time guard. This was flagged in the prior review and remains open.

**2. `len * 2` can silently wrap to 0** (Low, oracle_provider/src/lib.rs:350)

If an oracle response ever has `len > usize::MAX / 2` elements, `len * 2` wraps to a small value and `u32::try_from` succeeds — recording incorrect witness data without panicking. Unreachable in practice (gas budget caps response sizes), but `len.checked_mul(2).expect(...)` would make the intent explicit. This was also flagged in the prior review.

**3. `arg.a_len * 4` wrapping multiplication** (Low, callable_oracles/src/arithmetic/mod.rs:161)

Same pattern: `u64` multiplication without overflow check. In practice `a_len` comes from `BigintRepr.digits` which is bounded by gas-metered EVM input parsing, but `arg.a_len.checked_mul(4).unwrap()` would be strictly correct.

### Verified Correct (via subagent deep analysis)

- **Oracle query ordering preserved**: `DISCONNECT_ORACLE_QUERY_ID` correctly moved from `proof_running_system` (after `run_prepared`) into shared `post_op` functions in `basic_bootloader`. Both native and RISC-V modes now issue DISCONNECT at the same point.
- **Witness recording format**: `ReadWitnessSource` IOOracle impl produces a u32 stream byte-for-byte identical to the RISC-V simulator path — length prefix `len * 2`, LE word splitting confirmed by tests.
- **Arithmetic query packing**: Header `(q_len | r_len << 32)` round-trips correctly on both 32-bit and 64-bit consumers. Odd-count assertion guards against malformed oracle responses.
- **`USE_ADVICE` const generic**: `ForwardSystemTypes<O, true>` resolves to `NoStdSystemFunctions<true>`, matching `proof_running_system`. Both modes use oracle delegation for ecrecover/modexp.
- **`BasicBootloaderProvingExecutionConfig`**: Native prover input run uses the same execution config as RISC-V proving. Same `BlockDataKeeper`, `PostTxLoopOp`, and storage model parameters.
- **`bigint_op_delegation_with_carry_bit` cfg removal**: Correct and necessary — the native prover input path now exercises this code outside of tests.
- **Cross-mode equivalence testing**: `assert_block_outputs_match` thoroughly validates header, storage writes, account diffs, tx results, gas/native usage. `assert_eq!(prover_input_forward, proof_input)` verifies witness-level equivalence.
- **`compile_error!` guards**: Prevent compilation on unsupported (non-64-bit-LE) targets.
- **`write_bigint` MaybeUninit safety**: All digits fully initialized before `set_num_digits` — verified for both production and naive-advisor paths.

### Suggested Tests

- **Edge case: modexp with zero-length quotient on 64-bit path** — verify `q_len / 2 = 0` produces correct empty `BigintRepr` when dividend < modulus.
- **Property test: witness equivalence for transactions exercising all three oracle types** (arithmetic, blob KZG, field ops) — the current test only covers simple transactions. A test with a contract calling the modexp precompile would exercise the full native oracle path end-to-end.

File: callable_oracles/src/lib.rs [hunk 18]
```diff
@@ -60,12 +66,52 @@ impl UsizeDeserializable for MemoryRegionDescriptionParams {
...
+    let len_bytes = checked_byte_len(len_words_u8, core::mem::size_of::<u8>());
+
+    // Safety: the caller supplied a non-null pointer, and `len_bytes` was
+    // derived from a checked byte-length computation.
+    let bytes = unsafe { core::slice::from_raw_parts(ptr, len_bytes) };
+    bytes.to_vec()
+}
+
+#[inline(always)]
+pub(crate) fn read_host_struct<T>(ptr_u64: u64) -> T {
```
Comment (vv-dev-ai) [inline callable_oracles/src/lib.rs]: [AI-review] **Medium:** `read_host_struct<T>` has no trait bounds on `T`. Since `ptr::read()` does a bitwise copy, passing a type with drop glue would create a second owner — double-free UB. All current callers use `Copy` types, so this is sound today, but the API permits misuse.

Fix: `pub(crate) fn read_host_struct<T: Copy>(ptr_u64: u64) -> T`

Also, the safety comment should document the full precondition: the pointed-to memory must be initialized and valid for `size_of::<T>()` bytes within the current process's address space.

File: callable_oracles/src/arithmetic/mod.rs [hunk 19]
```diff
@@ -58,37 +106,265 @@ impl<M: MemorySource> OracleQueryProcessor<M> for ArithmeticQuery<M> {
...
+        let arg: ModExpAdviceParams64 = read_host_struct(arg_ptr as u64);
+
+        assert!(arg.a_ptr > 0);
+        assert!(arg.a_len > 0);
+        assert_eq!(arg.b_ptr, 0);
+        assert_eq!(arg.b_len, 0);
+        assert!(arg.modulus_ptr > 0);
+        assert!(arg.modulus_len > 0);
+
+        let a_len_u64_words = arg.a_len * 4;
```
Comment (vv-dev-ai) [inline callable_oracles/src/arithmetic/mod.rs]: [AI-review] **Low:** `arg.a_len * 4` and `arg.modulus_len * 4` (line 162) use wrapping `u64` multiplication. While unreachable in practice (gas metering bounds `a_len`), consider `arg.a_len.checked_mul(4).expect("a_len overflow")` for defensive correctness.

File: oracle_provider/src/lib.rs (hunk 11 above)
Comment (vv-dev-ai) [inline oracle_provider/src/lib.rs]: [AI-review] **Low:** `len * 2` can silently wrap to 0 if `len > usize::MAX / 2`, causing `u32::try_from` to succeed with an incorrect value. Unreachable under current gas budgets, but `len.checked_mul(2).expect("response length overflow")` makes the intent explicit and prevents silent data corruption.

File: callable_oracles/src/lib.rs (hunk 17 above)
Comment (antoniolocascio) [reply callable_oracles/src/lib.rs:75]: Added some doc, but the tldr is to validate non-null and alignment. In theory not needed, but just to harden the implementation

File: basic_bootloader/src/bootloader/block_flow/ethereum/post_tx_op_proving.rs (hunk 16 above)
Comment (antoniolocascio) [reply basic_bootloader/src/bootloader/block_flow/ethereum/post_tx_op_proving.rs]: Good point, done

**Comment by vv-dev-ai** [comment]
[AI-review] Reviewed commit `088e1424017a144a52b90b5069c3991dba9d7b45` — **APPROVE** (0 inline comments).

**Review by vv-dev-ai — APPROVED** [review]
[AI-review] PR #418 — **APPROVE**

Reply-review: re-evaluated the 3 open issues from the prior review against the latest commit.

- ✅ **RESOLVED** `readhoststructt-has-no-trait-bounds`: `read_host_struct<T: Copy>` bound added. Safety comment now documents all preconditions (non-null, alignment, initialization, liveness, no concurrent mutation).
- ✅ **RESOLVED** `len-times-2-overflow`: `len * 2` replaced with `len.checked_mul(2).expect("response length overflow")` in `oracle_provider/src/lib.rs:350`.
- ✅ **RESOLVED** `a-len-times-4-overflow`: Both `a_len` and `modulus_len` multiplications now use `.checked_mul(4).expect(...)` in `callable_oracles/src/arithmetic/mod.rs:161-165`.

File: callable_oracles/src/arithmetic/mod.rs (hunk 19 above)
Comment (antoniolocascio) [reply callable_oracles/src/arithmetic/mod.rs]: Done

File: oracle_provider/src/lib.rs (hunk 11 above)
Comment (antoniolocascio) [reply oracle_provider/src/lib.rs]: Done

**Review by AntonD3 — APPROVED** [review]
